lib/web/apps.go

```diff
@@ -0,0 +1,344 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package web implements web proxy handler that provides
+// web interface to view and connect to teleport nodes
+package web
+
+import (
+	"context"
+	"math/rand"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/events"
+	"github.com/gravitational/teleport/lib/httplib"
+	"github.com/gravitational/teleport/lib/reversetunnel"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/tlsca"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/web/ui"
+
+	"github.com/gravitational/trace"
+
+	"github.com/julienschmidt/httprouter"
+	log "github.com/sirupsen/logrus"
+)
+
+func (h *Handler) siteAppsGet(w http.ResponseWriter, r *http.Request, p httprouter.Params, ctx *SessionContext, site reversetunnel.RemoteSite) (interface{}, error) {
+	appClusterName := p.ByName("site")
+
+	// Get a list of application servers.
+	clt, err := ctx.GetUserClient(site)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	appServers, err := clt.GetAppServers(r.Context(), defaults.Namespace)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Get the public address of the proxy and remove the port.
+	proxyHost, err := utils.Host(h.cfg.ProxySettings.SSH.PublicAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return makeResponse(ui.MakeApps(h.auth.clusterName, proxyHost, appClusterName, appServers))
+}
+
+type createAppSessionRequest struct {
+	// FQDN is the full qualified domain name of the application.
+	FQDN string `json:"fqdn"`
+
+	// PublicAddr is the public address the application.
+	PublicAddr string `json:"public_addr"`
+
+	// ClusterName is the cluster within which this application is running.
+	ClusterName string `json:"cluster_name"`
+}
+
+type createAppSessionResponse struct {
+	// CookieValue is the application session cookie value.
+	CookieValue string `json:"value"`
+	// FQDN is application FQDN.
+	FQDN string `json:"fqdn"`
+}
+
+func (h *Handler) createAppSession(w http.ResponseWriter, r *http.Request, p httprouter.Params, ctx *SessionContext) (interface{}, error) {
```

Comment:
We can nicely test this method by via "appAPIGetter" similarly to Nodes and Users.